backend-elasticsearch/src/main/java/org/hibernate/search/v6poc/backend/elasticsearch/document/model/dsl/impl/AbstractElasticsearchIndexSchemaObjectNodeBuilder.java

```diff
@@ -20,37 +22,44 @@
 import org.hibernate.search.v6poc.backend.elasticsearch.logging.impl.Log;
 import org.hibernate.search.v6poc.util.impl.common.LoggerFactory;
 
-/**
- * A schema node builder.
- */
-abstract class AbstractIndexSchemaObjectNodeBuilder {
+abstract class AbstractElasticsearchIndexSchemaObjectNodeBuilder implements IndexSchemaObjectNodeBuilder {
 	private static final Log log = LoggerFactory.make( Log.class, MethodHandles.lookup() );
 
-	protected final DeferredInitializationIndexObjectFieldAccessor accessor =
-			new DeferredInitializationIndexObjectFieldAccessor();
-
 	private final Map<String, ElasticsearchIndexSchemaNodeContributor<PropertyMapping>> content = new HashMap<>();
 
 	@Override
 	public String toString() {
 		return new StringBuilder( getClass().getSimpleName() )
 				.append( "[" )
-				.append( ",absolutePath=" ).append( getAbsolutePath() )
+				.append( "absolutePath=" ).append( getAbsolutePath() )
 				.append( "]" )
 				.toString();
 	}
 
-	public IndexObjectFieldAccessor getAccessor() {
-		return accessor;
+	@Override
+	public IndexSchemaFieldContext addField(String relativeFieldName) {
+		ElasticsearchIndexSchemaFieldContextImpl fieldContext =
+				new ElasticsearchIndexSchemaFieldContextImpl( relativeFieldName );
+		putProperty( relativeFieldName, fieldContext );
+		return fieldContext;
 	}
 
-	public abstract String getAbsolutePath();
+	@Override
+	public IndexSchemaFieldContext createExcludedField(String relativeFieldName) {
+		return new ElasticsearchIndexSchemaFieldContextImpl( relativeFieldName );
+	}
 
-	public void putProperty(String name, ElasticsearchIndexSchemaNodeContributor<PropertyMapping> contributor) {
-		Object previous = content.putIfAbsent( name, contributor );
-		if ( previous != null ) {
-			throw log.indexSchemaNodeNameConflict( getAbsolutePath(), name);
-		}
+	@Override
+	public IndexSchemaObjectFieldNodeBuilder addObjectField(String relativeFieldName, ObjectFieldStorage storage) {
+		ElasticsearchIndexSchemaObjectFieldNodeBuilder objectFieldBuilder =
+				new ElasticsearchIndexSchemaObjectFieldNodeBuilder( getAbsolutePath(), relativeFieldName, storage );
+		putProperty( relativeFieldName, objectFieldBuilder );
+		return objectFieldBuilder;
+	}
+
+	@Override
+	public IndexSchemaObjectFieldNodeBuilder createExcludedObjectField(String relativeFieldName, ObjectFieldStorage storage) {
+		return new ElasticsearchIndexSchemaObjectFieldNodeBuilder( getAbsolutePath(), relativeFieldName, storage );
```

Comment:
Thinking out loud, not sure it's doable but I was wondering if:
- we could have `createField()`/`createObjectField()` methods (which would be used for exclusion or add - but maybe you expect the exclusion code to be different at some point?)
- have `addField()`/`addObjectField()` only do the add part
- have the ^ methods return true/false depending of if successful or not
- have the log call in the engine

Not totally sure it's a good idea but wanted to share.

Reply:
It was about the log call but also about improving the symmetry between the excluded fields and the added fields.

We can improve that later though if we think we have to. I'll merge this one.